src/parser/wakefiles.cpp

```diff
@@ -170,8 +171,33 @@ bool push_packaged_stdlib_files(std::vector<std::string> &out, const re2::RE2 &r
 
 #else
 
+struct profile_data {
+  std::chrono::time_point<std::chrono::steady_clock> start{std::chrono::steady_clock::now()};
+  size_t explored{0};
+  bool alerted_user{false};
+};
+
+static bool push_files_should_recurse(int dirfd, const std::string &path, struct dirent *f,
+                                      bool *failed) {
+  struct stat sbuf;
+
+#if defined(DT_DIR)
+  if (f->d_type != DT_UNKNOWN) {
+    return f->d_type == DT_DIR;
+  }
+#endif
+
+  if (fstatat(dirfd, f->d_name, &sbuf, AT_SYMLINK_NOFOLLOW) != 0) {
+    fprintf(stderr, "Failed to fstatat %s/%s: %s\n", path.c_str(), f->d_name, strerror(errno));
+    *failed = true;
+    return false;
+  }
+
+  return S_ISDIR(sbuf.st_mode);
+}
+
 static bool push_files(std::vector<std::string> &out, const std::string &path, int dirfd,
```

Comment:
I know you didn't write this but if you could explain what it does and what "push" means it would be helpful.